File: apps/libs/vending/softactivate/src/Licensing.Net/Validator.cs
//
// Copyright (c) 2014 SoftActivate. All rights reserved.
//

using System;
using System.Collections.Generic;
using System.Text; 

namespace SoftActivate.Licensing
{
    public class KeyValidator
    {
        public KeyValidator()
        {

        }

        public KeyValidator(LicenseTemplate licenseKeyTemplate)
            : this(licenseKeyTemplate, null)
        {

        }

        public KeyValidator(LicenseTemplate _keyTemplate, string _key)
        {
            SetKeyTemplate(_keyTemplate);
            if (_key != null)
                SetKey(_key);
        }

        ~KeyValidator()
        {

        }

        // Set the template used for validation (the template contains the license key format and the ECC verification key)
        public void SetKeyTemplate(LicenseTemplate templ)
        {
            if (templ == null)
                throw new Exception();

            keyTemplate = templ;
            validationData = new BitStream();
            validationData.Create(keyTemplate.ValidationDataSize);
        }

        // Set the license key used for validation and data query
        public void SetKey(string strKey)
        {
            string key = strKey;

            if (string.IsNullOrEmpty(strKey))
                throw new Exception("invalid license key");

            if (strKey.Length < keyTemplate.Header.Length +
                              keyTemplate.Footer.Length +
                              keyTemplate.CharactersPerGroup * keyTemplate.NumberOfGroups +
                              keyTemplate.GroupSeparator.Length * (keyTemplate.NumberOfGroups - 1))
            {
                throw new Exception("invalid license key (too short)");
            }

            // remove header and footer (if any)
            if (keyTemplate.Header != null)
                if (keyTemplate.Header.Length > 0)
                {
                    key = key.Remove(0, keyTemplate.Header.Length + 2);
                }

            if (keyTemplate.Footer != null)
                if (keyTemplate.Footer.Length > 0)
                {
                    key = key.Remove(key.Length - keyTemplate.Footer.Length - 2, keyTemplate.Footer.Length + 2);
                }

            // ungroup license key
            for (int i = 0, erasePos = 0; i < keyTemplate.NumberOfGroups - 1; i++)
            {
                erasePos += keyTemplate.CharactersPerGroup;
                key = key.Remove(erasePos, keyTemplate.GroupSeparator.Length);
            }

            rawKey = new BitStream();

            // decode license key
            switch (keyTemplate.KeyEncoding)
            {
                case LicenseKeyEncoding.Base32X:
                    {
                        byte[] keyBuf = Base32.Decode(key, 0, key.Length);
                        // reverse last byte
                        keyBuf[keyBuf.Length - 1] = (byte)(((keyBuf[keyBuf.Length - 1] * 0x0802LU & 0x22110LU) | (keyBuf[keyBuf.Length - 1] * 0x8020LU & 0x88440LU)) * 0x10101LU >> 16);

                        rawKey.Attach(keyBuf, (int)keyTemplate.KeyEncoding * keyTemplate.CharactersPerGroup * keyTemplate.NumberOfGroups);
                    }
                    break;

                case LicenseKeyEncoding.Base64:
                    {
                        if ((key.Length & 3) != 0)
                        {
                            key = key.PadRight(key.Length + 4 - (key.Length & 3), '=');
                        }

                        byte[] keyBuf = Convert.FromBase64String(key);
                        rawKey.Attach(keyBuf, (int)keyTemplate.KeyEncoding * keyTemplate.CharactersPerGroup * keyTemplate.NumberOfGroups);
                    }
                    break;

                default:
                    throw new Exception();
            }
        }

        public void SetValidationData(string fieldName, byte[] buf)
        {
            keyTemplate.validationDataMap.Attach(validationData);
            keyTemplate.validationDataMap.Set(fieldName, buf);
            keyTemplate.validationDataMap.Detach();
        }

        public void SetValidationData(string fieldName, int data)
        {
            keyTemplate.validationDataMap.Attach(validationData);
            keyTemplate.validationDataMap.Set(fieldName, data);
            keyTemplate.validationDataMap.Detach();
        }

        public void SetValidationData(string fieldName, string data)
        {
            keyTemplate.validationDataMap.Attach(validationData);
            keyTemplate.validationDataMap.Set(fieldName, data);
            keyTemplate.validationDataMap.Detach();
        }

        public void SetValidationData(string fieldName, DateTime date)
        {
            keyTemplate.validationDataMap.Attach(validationData);
            keyTemplate.validationDataMap.Set(fieldName, date);
            keyTemplate.validationDataMap.Detach();
        }

        // Verifies the license key's signature using the public(verification) ECC key from the template
        public bool IsKeyValid()
        {
            Verifier verifier = new Verifier();
            BitStream signedData = new BitStream(),
                      signature = new BitStream();

            signedData.Create(((keyTemplate.DataSize + keyTemplate.ValidationDataSize + 7) >> 3) << 3);
            signedData.Clear();
            if (keyTemplate.DataSize > 0)
                signedData.Write(rawKey.Buffer, keyTemplate.DataSize);
            if (keyTemplate.ValidationDataSize > 0)
                signedData.Write(validationData.Buffer, keyTemplate.ValidationDataSize);

            signedData.ZeroPadToNextByte();

            byte[] sigBytes = new byte[(keyTemplate.SignatureSize + 7) >> 3];
            rawKey.Seek(keyTemplate.DataSize);
            rawKey.Read(sigBytes, keyTemplate.SignatureSize);

            signature.Attach(sigBytes, keyTemplate.SignatureSize);
            signature.Seek(keyTemplate.SignatureSize);
            signature.ZeroPadToNextByte();

            // we use a different algorithm than ECDSA when the signature size must be smaller than twice the key size
            if (keyTemplate.SignatureSize < (keyTemplate.SignatureKeySize << 1))
                verifier.SetHashSize((uint)(keyTemplate.SignatureSize - keyTemplate.SignatureKeySize));
            else
                verifier.SetHashSize(0);

            verifier.SetPublicKey(keyTemplate.VerificationKey);

            return verifier.Verify(signedData.Buffer, 0, signedData.Buffer.Length, signature.Buffer, 0, signature.Buffer.Length, keyTemplate.SignatureSize);
        }

        // Queries the license key for data. The data is returned in raw format (a byte buffer).
        public byte[] QueryKeyData(string fieldName)
        {
            keyTemplate.keyDataMap.Attach(rawKey);
            byte[] result = keyTemplate.keyDataMap.Get(fieldName);
            keyTemplate.keyDataMap.Detach();
            return result;
        }

        public int QueryIntKeyData(string fieldName)
        {
            keyTemplate.keyDataMap.Attach(rawKey);
            int result = keyTemplate.keyDataMap.GetInt(fieldName);
            keyTemplate.keyDataMap.Detach();
            return result;
        }

        public DateTime QueryDateKeyData(string fieldName)
        {
            keyTemplate.keyDataMap.Attach(rawKey);
            DateTime result = keyTemplate.keyDataMap.GetDate(fieldName);
            keyTemplate.keyDataMap.Detach();
            return result;
        }

        public byte[] QueryValidationData(string fieldName)
        {
            keyTemplate.validationDataMap.Attach(validationData);
            byte[] result = keyTemplate.validationDataMap.Get(fieldName);
            keyTemplate.validationDataMap.Detach();
            return result;
        }

        private BitStream rawKey;
        private BitStream validationData;
        private LicenseTemplate keyTemplate;
    }
}

File: apps/libs/vending/softactivate/src/Licensing/validator.cpp
//
// Copyright (c) 2014 FreeGrabApp. All rights reserved.
//

#include "precomp.h"
#include "licensing.h"
#include "template.h"
#include "bitstream.h"
#include "base32.h"
#include "base64.h"
#include "sha1.h"
#include "except.h"
#include "uniconv.h"
#include "validator.h"

namespace FreeGrabApp {
    namespace Licensing {
        template<>
        KeyValidatorT<char>::KeyValidatorT():
        m_Impl( *new KeyValidatorImpl() )
        {

        }

        template<>
        KeyValidatorT<wchar_t>::KeyValidatorT():
        m_Impl( *new KeyValidatorImpl() )
        {

        }

        template<>
        KeyValidatorT<char>::KeyValidatorT(const LicenseTemplateT<char> * keyTemplate):
        m_Impl( *new KeyValidatorImpl(&keyTemplate->m_Impl) )
        {

        }

        template<>
        KeyValidatorT<wchar_t>::KeyValidatorT(const LicenseTemplateT<wchar_t> * keyTemplate):
        m_Impl( *new KeyValidatorImpl(&keyTemplate->m_Impl) )
        {

        }

        template<>
        KeyValidatorT<char>::~KeyValidatorT()
        {
            delete & m_Impl;
        }

        template<>
        KeyValidatorT<wchar_t>::~KeyValidatorT()
        {
            delete & m_Impl;
        }

        template<>
        KeyValidatorT<char> * KeyValidatorT<char>::Create()
        {
            return new KeyValidatorT<char>();
        }

        template<>
        KeyValidatorT<wchar_t> * KeyValidatorT<wchar_t>::Create()
        {
            return new KeyValidatorT<wchar_t>();
        }

        template<>
        void KeyValidatorT<char>::Destroy(KeyValidatorT<char> * obj)
        {
            delete obj;
        }

        template<>
        void KeyValidatorT<wchar_t>::Destroy(KeyValidatorT<wchar_t> * obj)
        {
            delete obj;
        }

        template<>
        void KeyValidatorT<char>::SetKeyTemplate(const LicenseTemplateT<char> * templ)
        {
            m_Impl.SetKeyTemplate(&templ->m_Impl);
        }

        template<>
        void KeyValidatorT<wchar_t>::SetKeyTemplate(const LicenseTemplateT<wchar_t> * templ)
        {
            m_Impl.SetKeyTemplate(&templ->m_Impl);
        }

        template<>
        void KeyValidatorT<char>::SetValidationData(const char * fieldName, const void * buf, int len)
        {
            USES_CONVERSION;
            m_Impl.SetValidationData(A2T(fieldName), buf, len);
        }

        template<>
        void KeyValidatorT<wchar_t>::SetValidationData(const wchar_t * fieldName, const void * buf, int len)
        {
        #ifndef _UNICODE
            USES_CONVERSION;
        #endif
            m_Impl.SetValidationData(W2T(fieldName), buf, len);
        }

        template<>
        void KeyValidatorT<char>::SetValidationData(const char * fieldName, const char * data)
        {
            USES_CONVERSION;
            m_Impl.SetValidationData(A2T(fieldName), A2T(data));
        }

        template<>
        void KeyValidatorT<wchar_t>::SetValidationData(const wchar_t * fieldName, const wchar_t * data)
        {
            USES_CONVERSION;
            m_Impl.SetValidationData(W2T(fieldName), W2T(data));
        }

        template<>
        void KeyValidatorT<char>::SetValidationData(const char * fieldName, int data)
        {
            USES_CONVERSION;
            m_Impl.SetValidationData(A2T(fieldName), data);
        }

        template<>
        void KeyValidatorT<wchar_t>::SetValidationData(const wchar_t * fieldName, int data)
        {
            USES_CONVERSION;
            m_Impl.SetValidationData(W2T(fieldName), data);
        }

        template<>
        void KeyValidatorT<char>::SetKey(const char * key)
        {
            USES_CONVERSION;
            m_Impl.SetKey(A2T(key));
        }

        template<>
        void KeyValidatorT<wchar_t>::SetKey(const wchar_t * key)
        {
            USES_CONVERSION;
            m_Impl.SetKey(W2T(key));
        }

        template<>
        bool KeyValidatorT<char>::IsKeyValid()
        {
            return m_Impl.IsKeyValid();
        }

        template<>
        bool KeyValidatorT<wchar_t>::IsKeyValid()
        {
            return m_Impl.IsKeyValid();
        }

        template<>
        void KeyValidatorT<char>::QueryKeyData(const char * dataField, void * buf, int * len)
        {
            USES_CONVERSION;
            m_Impl.QueryKeyData(A2T(dataField), buf, len);
        }

        template<>
        void KeyValidatorT<wchar_t>::QueryKeyData(const wchar_t * dataField, void * buf, int * len)
        {
            USES_CONVERSION;
            m_Impl.QueryKeyData(W2T(dataField), buf, len);
        }

        template<>
        int KeyValidatorT<char>::QueryIntKeyData(const char * dataField)
        {
            USES_CONVERSION;
            return m_Impl.QueryIntKeyData(A2T(dataField));
        }

        template<>
        int KeyValidatorT<wchar_t>::QueryIntKeyData(const wchar_t * dataField)
        {
            USES_CONVERSION;
            return m_Impl.QueryIntKeyData(W2T(dataField));
        }

        template<>
        void KeyValidatorT<char>::QueryDateKeyData(const char * dataField, int * year, int * month, int * day)
        {
            USES_CONVERSION;
            m_Impl.QueryDateKeyData(A2T(dataField), year, month, day);
        }

        template<>
        void KeyValidatorT<wchar_t>::QueryDateKeyData(const wchar_t * dataField, int * year, int * month, int * day)
        {
            USES_CONVERSION;
            m_Impl.QueryDateKeyData(W2T(dataField), year, month, day);
        }

        template<>
        void KeyValidatorT<char>::QueryValidationData(const char * dataField, void * buf, int * len)
        {
            USES_CONVERSION;
            m_Impl.QueryValidationData((dataField) ? A2T(dataField) : NULL, buf, len);
        }

        template<>
        void KeyValidatorT<wchar_t>::QueryValidationData(const wchar_t * dataField, void * buf, int * len)
        {
            USES_CONVERSION;
            m_Impl.QueryValidationData((dataField) ? W2T(dataField) : NULL, buf, len);
        }
    };
};
